### PR TITLE
feat(common): make date/number formatting functions public

### PR DESCRIPTION
**Pull request by ocombe**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

If you want to use the number, percent, currency and date pipes outside of your templates, you need to instantiate those pipes and manually transform the values

Issue Number: #20536

## What is the new behavior?
The utility functions `formatNumber`, `formatPercent`, `formatCurrency`, and `formatDate` used by the number, percent, currency and date pipes are now available for developers who want to use them outside of templates.

## Does this PR introduce a breaking change?
```
[x] No
```

### CODE REVIEW

File: packages/common/src/i18n/format_date.ts [hunk 1]
```diff
@@ -40,25 +43,26 @@ enum TranslationType {
 /**
  * Transforms a date to a locale string based on a pattern and a timezone
  *
- * @internal
+ * @stable
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: needs more doc !

File: packages/common/src/i18n/format_date.ts [hunk 2]
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
   const timezoneOffset = timezoneToOffset(timezone, dateTimezoneOffset);
   return addDateMinutes(date, reverseValue * (timezoneOffset - dateTimezoneOffset));
 }
+
+/**
+ * Transforms a value into a date
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: comment here do no say anything more than `toDate()`. 
Pls improve

Comment (vicb) [reply packages/common/src/i18n/format_date.ts]: refine type also

File: packages/common/src/i18n/format_date.ts [hunk 3]
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+  }
+
+  let match: RegExpMatchArray|null;
+  if (isDate(value)) {
+    return value;
+  } else if (!isNaN(value - parseFloat(value))) {
+    return new Date(parseFloat(value));
+  } else if (typeof value === 'string' && /^(\d{4}-\d{1,2}-\d{1,2})$/.test(value)) {
+    /**
+     * For ISO Strings without time the day, month and year must be extracted from the ISO String
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: move to API doc of the function ?

File: packages/common/src/i18n/format_date.ts [hunk 4]
```diff
@@ -607,3 +613,66 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+  }
+
+  const date = new Date(value);
+  if (!isDate(date)) {
+    throw new Error(`Unable to transform "${value}" into a date`);
+  }
+  return date;
+}
+
+export function isoStringToDate(match: RegExpMatchArray): Date {
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: add a comment here pls

File: packages/common/src/i18n/format_number.ts [hunk 5]
```diff
@@ -140,7 +140,7 @@ function formatNumber(
 /**
  * Formats a currency to a locale string
  *
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: please add docs

File: packages/common/src/i18n/format_number.ts [hunk 6]
```diff
@@ -162,12 +162,12 @@ export function formatCurrency(
 /**
  * Formats a percentage to a locale string
  *
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: please add docs

File: packages/common/src/pipes/date_pipe.ts [hunk 7]
```diff
@@ -120,6 +116,9 @@ export const ISO8601_DATE_REGEX =
  *   pipe needs to re-run (this is to avoid reformatting the date on every change detection run
  *   which would be an expensive operation).
  *
+ * If you want to use this pipe outside of a template, you can use the function
```
Comment (vicb) [inline packages/common/src/pipes/date_pipe.ts]: no you don't want to use the pipe outside of a template. Remove ?

File: packages/common/src/pipes/number_pipe.ts [hunk 8]
```diff
@@ -31,6 +31,9 @@ import {invalidPipeArgumentError} from './invalid_pipe_argument_error';
  * For more information on the acceptable range for each of these numbers and other
  * details see your native internationalization library.
  *
+ * If you want to use this pipe outside of a template, you can use the function
```
Comment (vicb) [inline packages/common/src/pipes/number_pipe.ts]: same comment as other non-pipes

File: packages/common/test/pipes/date_pipe_spec.ts [hunk 9]
```diff
@@ -47,28 +48,44 @@ import localeAr from '@angular/common/locales/ar';
     });
 
     describe('supports', () => {
-      it('should support date', () => { expect(() => pipe.transform(date)).not.toThrow(); });
+      it('should support date', () => {
+        expect(isDate(toDate(date))).toBeTruthy();
```
Comment (vicb) [inline packages/common/test/pipes/date_pipe_spec.ts]: what is the intent here ?
Does this belong to `date_pipe_spec.ts` ?

**Review by vicb — CHANGES_REQUESTED** [review]
Please
- Add or expand the API docs. Especially important for API docs. Internal functions should also have a short comment,
- refine date type from "any" to a|b|c,
- I was expected to see more tests moved to the new format spec. It doesn't make sense any more to have them in the pipes if pipes only forward to those functions (only a sanity check is ok for the pipes).

File: packages/common/src/i18n/format_date.ts (hunk 3 above)
Comment (ocombe) [reply packages/common/src/i18n/format_date.ts]: No, it's there to explain the method used to convert the ISO string to a date, but it has no value for people who use the function.

File: packages/common/src/i18n/format_date.ts [hunk 10]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
 
 /**
- * Transforms a date to a locale string based on a pattern and a timezone
+ * @ngModule CommonModule
+ * @whatItDoes Formats a date according to locale rules.
+ * @description
  *
- * @internal
+ * Where:
+ * - `value` is a date object or a number (milliseconds since UTC epoch) or an ISO string
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: please update the type accordingly

Comment (vicb) [reply packages/common/src/i18n/format_date.ts]: "is a `Date` or"

File: packages/common/src/i18n/format_date.ts [hunk 11]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ * (https://www.w3.org/TR/NOTE-datetime).
+ * - `format` indicates which date/time components to include. See {@link DatePipe} for a detailed
+ * description.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: no default

File: packages/common/src/i18n/format_date.ts [hunk 12]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ * default).
+ * - `timezone` to be used for formatting. It understands UTC/GMT and the continental US time zone
+ *  abbreviations, but for general use, use a time zone offset, for example,
+ *  `'+0430'` (4 hours, 30 minutes east of the Greenwich meridian)
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: remove "(4 hours, 30 minutes east of the Greenwich meridian)" - no added value and technically wrong GMT vs UTC

File: packages/common/src/i18n/format_date.ts [hunk 13]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ *  If not specified, the local system timezone of the end-user's browser will be used.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: "the local system timezone of the end-user's browser will be used." -> host system settings are used.

File: packages/common/src/i18n/format_date.ts [hunk 14]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ *
+ * See {@link DatePipe} for more details about the parameters.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: See {@link DatePipe} for more details

File: packages/common/src/i18n/format_date.ts [hunk 15]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ *
+ * @stable
  */
-export function formatDate(date: Date, format: string, locale: string, timezone?: string): string {
+export function formatDate(value: any, format: string, locale: string, timezone?: string): string {
+  let date = toDate(value);
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: const

File: packages/common/src/i18n/format_date.ts [hunk 16]
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+    const parsedNb = parseFloat(value);
+
+    // any string that only contains numbers, like "1234" but not like "1234hello"
+    if (!isNaN(value as any - parsedNb)) {
+      return new Date(parsedNb);
+    }
+
+    if (/^(\d{4}-\d{1,2}-\d{1,2})$/.test(value)) {
+      /**
+       * For ISO Strings without time the day, month and year must be extracted from the ISO String
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: move to the API doc of toDate

File: packages/common/src/i18n/format_date.ts [hunk 17]
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
   const timezoneOffset = timezoneToOffset(timezone, dateTimezoneOffset);
   return addDateMinutes(date, reverseValue * (timezoneOffset - dateTimezoneOffset));
 }
+
+/**
+ * Transforms a value into a date. The supported values are dates, numbers, numeric strings
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/common/src/i18n/format_date.ts]: throw if not able to convert to a date

File: packages/common/src/i18n/format_date.ts [hunk 18]
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+
+  const date = new Date(value as any);
+  if (!isDate(date)) {
+    throw new Error(`Unable to transform "${value}" into a date`);
+  }
+  return date;
+}
+
+/**
+ * This function converts a string matching the ISO8601_DATE_REGEX to a date.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: "This function converts a string matching the ISO8601_DATE_REGEX to a date." 
What does this means ?
What about "Convert a date in ISO8601 to a `Date`" ?

File: packages/common/src/i18n/format_date.ts [hunk 19]
```diff
@@ -607,3 +629,86 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+  const dateSetter = match[8] ? date.setUTCFullYear : date.setFullYear;
+  const timeSetter = match[8] ? date.setUTCHours : date.setHours;
+
+  // if there is a timezone defined like "+01:00" or "+0100"
+  if (match[9]) {
+    tzHour = +(match[9] + match[10]);
+    tzMin = +(match[9] + match[11]);
+  }
+  dateSetter.call(date, +(match[1]), +(match[2]) - 1, +(match[3]));
+  const h = +(match[4] || '0') - tzHour;
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: + -> Number, '0' -> 0

File: packages/common/src/i18n/format_number.ts [hunk 20]
```diff
@@ -35,7 +35,7 @@ function strToNumber(value: number | string): number {
 /**
  * Transforms a number to a locale string based on a style and a format
  */
-function formatNumber(
+function formatNbToStr(
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: formatNbAsString ?

File: packages/common/src/i18n/format_number.ts [hunk 21]
```diff
@@ -138,9 +138,22 @@ function formatNumber(
...
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number as currency using locale rules.
+ * @description
  *
- * @internal
+ * Use `currency` to format a number as currency.
+ *
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: no default

File: packages/common/src/i18n/format_date.ts [hunk 22]
```diff
@@ -38,11 +41,28 @@ enum TranslationType {
...
+ * @ngModule CommonModule
+ * @whatItDoes Formats a date according to locale rules.
+ * @description
  *
- * @internal
+ * Where:
+ * - `value` is a date object or a number (milliseconds since UTC epoch) or an ISO string
+ * (https://www.w3.org/TR/NOTE-datetime).
+ * - `format` indicates which date/time components to include. See {@link DatePipe} for a detailed
+ * description.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: nit: I think aligning the text for continued line is easier to read

File: packages/common/src/i18n/format_number.ts [hunk 23]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
- * Formats a percentage to a locale string
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number as a percentage according to locale rules.
+ * @description
+ *
+ * Formats a number as percentage.
  *
- * @internal
+ * Where:
+ * - `value` is a number.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: does not match the function !

File: packages/common/src/i18n/format_number.ts [hunk 24]
```diff
@@ -138,9 +138,22 @@ function formatNumber(
...
- * Formats a currency to a locale string
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number as currency using locale rules.
+ * @description
  *
- * @internal
+ * Use `currency` to format a number as currency.
+ *
+ * Where:
+ * - `value` is a number.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: does not match the fn - number only seems to be the right thing

File: packages/common/src/i18n/format_number.ts [hunk 25]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number as a percentage according to locale rules.
+ * @description
+ *
+ * Formats a number as percentage.
  *
- * @internal
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: no default

File: packages/common/src/i18n/format_number.ts [hunk 26]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
- * Formats a number to a locale string
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number according to locale rules.
+ * @description
+ *
+ * Formats a number as text. Group sizing and separator and other locale-specific
+ * configurations are based on the active locale.
+ *
+ * Where:
+ * - `value` is a number.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: again

File: packages/common/src/i18n/format_number.ts [hunk 27]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
 }
 
 /**
- * Formats a number to a locale string
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number according to locale rules.
+ * @description
+ *
+ * Formats a number as text. Group sizing and separator and other locale-specific
+ * configurations are based on the active locale.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: active locale -> locale.

File: packages/common/src/i18n/format_number.ts [hunk 28]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
+ *
+ * Where:
+ * - `value` is a number.
+ * - `locale` is a `string` defining the locale to use (uses the current {@link LOCALE_ID} by
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: no default

File: packages/common/src/i18n/format_number.ts [hunk 29]
```diff
@@ -160,28 +173,54 @@ export function formatCurrency(
...
+ * default).
+ * - `digitInfo` is a `string` which has a following format: <br>
+ *   <code>{minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}</code>.
+ *   - `minIntegerDigits` is the minimum number of integer digits to use. Defaults to `1`.
+ *   - `minFractionDigits` is the minimum number of digits after fraction. Defaults to `0`.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: "after fraction" is it the correct terminology ?

File: packages/common/test/i18n/format_date_spec.ts [hunk 30]
```diff
@@ -0,0 +1,317 @@
...
+
+      it('should throw for NaN', () => { expect(() => toDate(Number.NaN)).toThrow(); });
+
+      it('should support ISO string without time',
+         () => { expect(isDate(toDate('2015-01-01'))).toBeTruthy(); });
+
+      it('should throw for objects', () => { expect(() => toDate({} as any)).toThrow(); });
+    });
+
+    describe('transform', () => {
```
Comment (vicb) [inline packages/common/test/i18n/format_date_spec.ts]: transform ?

File: packages/common/test/i18n/format_number_spec.ts [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+ */
+
+import localeEn from '@angular/common/locales/en';
+import localeEsUS from '@angular/common/locales/es-US';
+import localeFr from '@angular/common/locales/fr';
+import localeAr from '@angular/common/locales/ar';
+import {formatCurrency, formatNumber, formatPercent, registerLocaleData} from '@angular/common';
+import {describe, expect, it} from '@angular/core/testing/src/testing_internal';
+
+{
```
Comment (vicb) [inline packages/common/test/i18n/format_number_spec.ts]: remove this

File: packages/common/test/i18n/format_date_spec.ts [hunk 32]
```diff
@@ -0,0 +1,317 @@
...
+import localeAr from '@angular/common/locales/ar';
+import localeDe from '@angular/common/locales/de';
+import localeEn from '@angular/common/locales/en';
+import localeEnExtra from '@angular/common/locales/extra/en';
+import localeHu from '@angular/common/locales/hu';
+import localeSr from '@angular/common/locales/sr';
+import localeTh from '@angular/common/locales/th';
+import {isDate, toDate, formatDate} from '@angular/common/src/i18n/format_date';
+
+{
```
Comment (vicb) [inline packages/common/test/i18n/format_date_spec.ts]: remove

File: packages/common/test/pipes/date_pipe_spec.ts [hunk 33]
```diff
@@ -58,262 +46,23 @@ import localeAr from '@angular/common/locales/ar';
          () => { expect(() => pipe.transform('123456789.11')).not.toThrow(); });
 
       it('should support ISO string',
-         () => expect(() => pipe.transform('2015-06-15T21:43:11Z')).not.toThrow());
+         () => { expect(() => pipe.transform('2015-06-15T21:43:11Z')).not.toThrow(); });
```
Comment (vicb) [inline packages/common/test/pipes/date_pipe_spec.ts]: revert the diffs ?

File: packages/common/src/i18n/format_date.ts (hunk 15 above)
Comment (ocombe) [reply packages/common/src/i18n/format_date.ts]: we change the value later if there is a timezone

File: packages/common/src/i18n/format_date.ts [hunk 34]
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
   const timezoneOffset = timezoneToOffset(timezone, dateTimezoneOffset);
   return addDateMinutes(date, reverseValue * (timezoneOffset - dateTimezoneOffset));
 }
+
+/**
+ * Converts a value to date
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: missing final .

File: packages/common/src/i18n/format_date.ts [hunk 35]
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+
+  const date = new Date(value as any);
+  if (!isDate(date)) {
+    throw new Error(`Unable to convert "${value}" into a date`);
+  }
+  return date;
+}
+
+/**
+ * Convert a date in ISO8601 to a Date.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: Convert*s*

File: packages/common/src/i18n/format_date.ts [hunk 36]
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
...
+ * Used instead of Date.parse because of browser discrepancies.
```
Comment (vicb) [inline packages/common/src/i18n/format_date.ts]: quote `Date.parse`

File: packages/common/src/i18n/format_number.ts [hunk 37]
```diff
@@ -138,9 +138,21 @@ function formatNumber(
...
- * Formats a currency to a locale string
+ * @ngModule CommonModule
+ * @whatItDoes Formats a number as currency using locale rules.
+ * @description
  *
- * @internal
+ * Use `currency` to format a number as currency.
+ *
+ * Where:
+ * - `value` is a number or numeric string.
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: wrong way ! -> number

File: packages/common/src/i18n/format_number.ts [hunk 38]
```diff
@@ -160,28 +172,52 @@ export function formatCurrency(
...
+ * @whatItDoes Formats a number according to locale rules.
+ * @description
+ *
+ * Formats a number as text. Group sizing and separator and other locale-specific
+ * configurations are based on the locale.
+ *
+ * Where:
+ * - `value` is a number or numeric string.
+ * - `locale` is a `string` defining the locale to use.
+ * - `digitInfo` is a `string` which has a following format: <br>
```
Comment (vicb) [inline packages/common/src/i18n/format_number.ts]: add to all pipes or no pipe (other have `{@link ...}`)

File: packages/common/src/i18n/format_date.ts [hunk 39]
```diff
@@ -165,8 +184,10 @@ function padNumber(
...
     }
   }
-  let strNum = '' + num;
-  while (strNum.length < digits) strNum = '0' + strNum;
+  let strNum = String(num);
+  while (strNum.length < digits) {
+    strNum = '0' + strNum;
+  }
   if (trim) {
     strNum = strNum.substr(strNum.length - digits);
```
Comment (wKoza) [inline packages/common/src/i18n/format_date.ts]: `strNum.length - digits` isn't always equal to `0`?

File: packages/common/src/i18n/format_date.ts [hunk 40]
```diff
@@ -607,3 +628,90 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
   const timezoneOffset = timezoneToOffset(timezone, dateTimezoneOffset);
   return addDateMinutes(date, reverseValue * (timezoneOffset - dateTimezoneOffset));
 }
+
+/**
+ * Converts a value to date
+ *
+ * Supported input formats:
+ * - `Date`s
```
Comment (wKoza) [inline packages/common/src/i18n/format_date.ts]: `Date`

File: packages/common/src/i18n/format_date.ts (hunk 39 above)
Comment (ocombe) [reply packages/common/src/i18n/format_date.ts]: not always, we can have additional zeros

File: packages/common/src/i18n/format_number.ts [hunk 41]
```diff
@@ -18,34 +18,19 @@ const DIGIT_CHAR = '#';
...
-  }
-  return value;
-}
-
 /**
  * Transforms a number to a locale string based on a style and a format
  */
-function formatNumber(
-    value: number | string, pattern: ParsedNumberFormat, locale: string, groupSymbol: NumberSymbol,
+function formatNbToLocaleString(
```
Comment (IgorMinar) [inline packages/common/src/i18n/format_number.ts]: please avoid using custom abbreviations in non-local variables. formatNbToLocaleString -> formatNumberToLocaleString

anyone not familiar with our codebase that sees "Nb" will have to question what it means. This makes our code base harder to read and change by people not familiar with every change we make - e.g. our community, but also team members working on other parts of the code base that come across this code.

**Review by IgorMinar — APPROVED** [review]
the rest looks good. thanks

**Review by vicb — APPROVED** [review]

**Comment by vicb** [comment]
[presubmit](	http://test/OCL:186822308:BASE:186822499:1519424521742:7ea2c4fa
)

**Pull request closed without merging**